File: short_questions/compny-coding-problems/golang/scenario_problems.go
package main

import (
	"fmt"
)

// 91. Exception Handling Logic (Scenario)
// In Go, we use panic/recover for exceptions
func exceptionHandlingDemo() {
	defer func() {
		if r := recover(); r != nil {
			fmt.Println("Recovered from exception:", r)
		}
	}()
	panic("Something went wrong!")
}

// 92. Method Overloading Logic (Not supported in Go directly, simulated)
// Go doesn't support method overloading. We use different names or interfaces.
func add(a, b int) int {
	return a + b
}
func addThree(a, b, c int) int {
	return a + b + c
}

// 93. Pass by Value vs Pass by Reference
func modifyValue(val int, ref *int) {
	val = 100
	*ref = 100
}

// 94. String Pool Logic (Go internals)
// Go strings are immutable and behave somewhat effectively, but no direct "String Pool" manipulation like Java.
func stringIdentity() {
	s1 := "Hello"
	s2 := "Hello"
	fmt.Println("s1 == s2:", s1 == s2) // True (Content comparison)
}

// 95. Switch Case Logic
func checkGrade(score int) {
	switch {
	case score >= 90:
		fmt.Println("A")
	case score >= 80:
		fmt.Println("B")
	case score >= 70:
		fmt.Println("C")
	default:
		fmt.Println("F")
	}
}

// 96. Static keyword logic (No static in Go, use package level vars)
var count int = 0

func counter() {
	count++
	fmt.Println("Count:", count)
}

// 97. Final/Const logic
func constantDemo() {
	const Pi = 3.14
	// Pi = 3.14159 // Error: cannot assign to Pi
	fmt.Println("Pi:", Pi)
}

// 98. Object Cloning (Struct copy)
type Person struct {
	Name string
	Age  int
}

func cloneDemo() {
	p1 := Person{"John", 30}
	p2 := p1 // Value copy (Shallow clone)
	p2.Name = "Doe"
	fmt.Println("p1:", p1) // John
	fmt.Println("p2:", p2) // Doe
}

// 99. Constructor Logic (Factory functions)
func NewPerson(name string, age int) Person {
	return Person{Name: name, Age: age}
}

// 100. Immutable Class Logic (Private fields with getters)
type ImmutablePoint struct {
	x, y int
}

func NewImmutablePoint(x, y int) ImmutablePoint {
	return ImmutablePoint{x: x, y: y}
}
func (p ImmutablePoint) X() int { return p.x }
func (p ImmutablePoint) Y() int { return p.y }

func main() {
	fmt.Println("91. Exception Handling:")
	exceptionHandlingDemo()

	fmt.Println("\n92. Overloading (Simulated):")
	fmt.Println("Add(1, 2):", add(1, 2))
	fmt.Println("AddThree(1, 2, 3):", addThree(1, 2, 3))

	fmt.Println("\n93. Pass by Value vs Ref:")
	v, r := 1, 1
	modifyValue(v, &r)
	fmt.Printf("Val: %d (Unchanged), Ref: %d (Changed)\n", v, r)

	fmt.Println("\n94. String Identity:")
	stringIdentity()

	fmt.Println("\n95. Switch Case:")
	checkGrade(85)

	fmt.Println("\n96. Static (Package Var):")
	counter()
	counter()

	fmt.Println("\n97. Constant:")
	constantDemo()

	fmt.Println("\n98. Cloning:")
	cloneDemo()

	fmt.Println("\n99. Constructor:")
	p := NewPerson("Alice", 25)
	fmt.Println(p)

	fmt.Println("\n100. Immutable Point:")
	ip := NewImmutablePoint(10, 20)
	fmt.Println(ip.X(), ip.Y())
}

File: short_questions/compny-coding-problems/java/ScenarioProblems.java
package java_solutions;

import java.util.*;

public class ScenarioProblems {

    // 91. Exception Handling Logic
    public static void exceptionHandlingDemo() {
        try {
            int result = 10 / 0;
        } catch (ArithmeticException e) {
            System.out.println("Caught exception: " + e.getMessage());
        } finally {
            System.out.println("Finally block executed");
        }
    }

    // 92. Method Overloading Logic
    public static int add(int a, int b) {
        return a + b;
    }

    public static int add(int a, int b, int c) {
        return a + b + c;
    }

    public static double add(double a, double b) {
        return a + b;
    }

    // 93. Pass by Value vs Pass by Reference (Java is always Pass by Value)
    public static void modifyValue(int val, StringBuilder ref) {
        val = 100;
        ref.append(" World");
    }

    // 94. String Pool Logic
    public static void stringIdentity() {
        String s1 = "Hello";
        String s2 = "Hello";
        String s3 = new String("Hello");
        System.out.println("s1 == s2: " + (s1 == s2)); // True (String Pool)
        System.out.println("s1 == s3: " + (s1 == s3)); // False (Heap)
        System.out.println("s1.equals(s3): " + s1.equals(s3)); // True (Content)
    }

    // 95. Switch Case Logic
    public static void checkGrade(int score) {
        switch (score / 10) {
            case 10:
            case 9:
                System.out.println("A");
                break;
            case 8:
                System.out.println("B");
                break;
            case 7:
                System.out.println("C");
                break;
            default:
                System.out.println("F");
        }
    }

    // 96. Static keyword logic
    static int count = 0;

    public static void counter() {
        count++;
        System.out.println("Count: " + count);
    }

    // 97. Final keyword logic
    public static void finalDemo() {
        final double PI = 3.14159;
        // PI = 3.14; // Compile Error
        System.out.println("PI: " + PI);
    }

    // 98. Object Cloning
    static class Person implements Cloneable {
        String name;
        int age;

        Person(String name, int age) {
            this.name = name;
            this.age = age;
        }

        @Override
        protected Object clone() throws CloneNotSupportedException {
            return super.clone();
        }

        @Override
        public String toString() {
            return name + " (" + age + ")";
        }
    }

    public static void cloneDemo() {
        try {
            Person p1 = new Person("John", 30);
            Person p2 = (Person) p1.clone();
            p2.name = "Doe";
            System.out.println("p1: " + p1); // John
            System.out.println("p2: " + p2); // Doe
        } catch (CloneNotSupportedException e) {
            e.printStackTrace();
        }
    }

    // 99. Constructor Logic
    static class Employee {
        String name;
        int id;

        Employee(String name, int id) {
            this.name = name;
            this.id = id;
        }

        // Constructor Chaining
        Employee(String name) {
            this(name, 0);
        }
    }

    // 100. Immutable Class Logic
    final static class ImmutablePoint {
        private final int x;
        private final int y;

        public ImmutablePoint(int x, int y) {
            this.x = x;
            this.y = y;
        }

        public int getX() {
            return x;
        }

        public int getY() {
            return y;
        }
    }

    public static void main(String[] args) {
        System.out.println("91. Exception Handling:");
        exceptionHandlingDemo();

        System.out.println("\n92. Overloading:");
        System.out.println("Add(1, 2): " + add(1, 2));
        System.out.println("Add(1.5, 2.5): " + add(1.5, 2.5));

        System.out.println("\n93. Pass by Value vs Ref:");
        int v = 1;
        StringBuilder r = new StringBuilder("Hello");
        modifyValue(v, r);
        System.out.println("Val: " + v + " (Unchanged), Ref: " + r + " (Changed content)");

        System.out.println("\n94. String Identity:");
        stringIdentity();

        System.out.println("\n95. Switch Case:");
        checkGrade(85);

        System.out.println("\n96. Static:");
        counter();
        counter();

        System.out.println("\n97. Final:");
        finalDemo();

        System.out.println("\n98. Cloning:");
        cloneDemo();

        System.out.println("\n99. Constructor:");
        Employee emp = new Employee("Alice");
        System.out.println("Emp: " + emp.name + ", " + emp.id);

        System.out.println("\n100. Immutable Point:");
        ImmutablePoint ip = new ImmutablePoint(10, 20);
        System.out.println("X=" + ip.getX() + ", Y=" + ip.getY());
    }
}
